packages/test-sync/src/commonTest/kotlin/io/realm/kotlin/test/mongodb/shared/SubscriptionExtensionsTests.kt

```diff
@@ -0,0 +1,343 @@
+        ).initialSubscriptions { realm: Realm ->
+            realm.query<FlexParentObject>("section = $0", section).subscribe()
+        }.build()
+
+        Realm.open(config).use { realmFromAnotherDevice ->
+            realmFromAnotherDevice.writeBlocking {
+                copyToRealm(FlexParentObject(section))
+            }
+            realmFromAnotherDevice.syncSession.uploadAllLocalChanges(30.seconds)
+        }
+
+        // Data still hasn't reached this device
+        assertEquals(0, realm.query<FlexParentObject>().count().find())
+        // Check that subscribing to a query, will run the query on the data downloaded from
+        // the server and not just local data, due to WaitForSync.FIRST_TIME being the default.
+        val result = realm.query<FlexParentObject>("section = $0", section).subscribe()
+        assertEquals(1, result.size)
+        assertEquals(section, result.first().section)
+    }
+
+    @Test
+    fun realmQuery_subscribe_waitFirstTime() = runBlocking<Unit> {
+        // Unnamed
+        realm.query<FlexParentObject>().subscribe() // Default value is WaitForSync.FIRST_TIME
+        var updatedSubs = realm.subscriptions
+        assertEquals(1, updatedSubs.size)
+        assertEquals(SubscriptionSetState.COMPLETE, updatedSubs.state)
+        var sub: Subscription = updatedSubs.first()
+        assertNull(sub.name)
+        assertEquals("TRUEPREDICATE ", sub.queryDescription)
+        assertEquals("FlexParentObject", sub.objectType)
+
+        // Named
+        realm.query<FlexParentObject>().subscribe("my-name") // Default value is WaitForSync.FIRST_TIME
+        updatedSubs = realm.subscriptions
+        assertEquals(2, updatedSubs.size)
+        assertEquals(SubscriptionSetState.COMPLETE, updatedSubs.state)
+        sub = updatedSubs.last()
+        assertEquals("my-name", sub.name)
+        assertEquals("TRUEPREDICATE ", sub.queryDescription)
+        assertEquals("FlexParentObject", sub.objectType)
+    }
+
+    @Test
+    fun realmQuery_subscribe_waitNever() = runBlocking {
+        // Un-named
+        realm.query<FlexParentObject>().subscribe(mode = WaitForSync.NEVER)
+        var updatedSubs = realm.subscriptions
+        assertEquals(1, updatedSubs.size)
+        // Updating the subscription will happen in the background, but
+        // hopefully hasn't reached COMPLETE yet.
+        assertNotEquals(SubscriptionSetState.COMPLETE, updatedSubs.state)
+
+        // Named
+        realm.query<FlexParentObject>().subscribe(name = "my-name", mode = WaitForSync.NEVER)
+        updatedSubs = realm.subscriptions
+        assertEquals(2, updatedSubs.size)
+        // Updating the subscription will happen in the background, but
+        // hopefully hasn't reached COMPLETE yet.
+        assertNotEquals(SubscriptionSetState.COMPLETE, updatedSubs.state)
+    }
+
+    @Test
+    fun realmQuery_subscribe_waitAlways() = runBlocking {
```

Comment:
Wouldn't it be safer to use the pattern from `realmQuery_subscribe_queryResultIsLatestVersion` to verify that you always get a result reflecting writes from another "device"?

Reply:
Similar to above, we have other tests showing that once a subscription set is COMPLETE, data is there.

Reply:
I don't get your comment. You are in fact asserting on data here 🤔 

My point was that it seemed safer to rely on writing data to another realm and asserting things after that was downloaded than relying on pause/resume, but 🤷

Reply:
Ah sorry, I was a bit fast there 🙈 ... Hmm, I'm not sure that uploading from another device is safer since we cannot control the timing of it, i.e. the data might already be downloaded as part of subscribing the first time, so we would be checking the wrong things:

```
      val results1 = realm.query<FlexParentObject>("section = $0", sectionId).subscribe(mode = WaitForSync.FIRST_TIME)
       uploadServerData(sectionId, 5)
      val results2 = realm.query<FlexParentObject>("section = $0", sectionId).subscribe(mode =  WaitForSync.ALWAYS)
```

The data uploaded would already be part of the subscription created first, so regardless of what the 2nd call does, the data might already be there. I agree the test is a bit awkward, but I don't see another way of doing it?

Reply:
If you do something like:
```
1. Subscribe to realm2 with some predicate A and check initial data
2. Write data that doesn't match predicate A but match predicate B to realm1
3. Update subscription with `ALWAYS` on realm2 to match predicate B
```
Wouldn't that ensure that data is always available when the subscription i 3. returns?

Reply:
The problem with 3) is that you cannot update the subscription to a new query, it would be considered "first time" that way.